### DISCUSSION TITLE
Event binding

### DISCUSSION THREAD

**Discussion by Umenokin** [AvaloniaUI/Avalonia#9079] [Q&A] [answered]
I am really struggling to find any information on command/event binding between `UserControl` and `ViewModel`

These are resources I was able to find about the subject:
https://docs.avaloniaui.net/guides/deep-dives/reactiveui/view-activation

## My assumptions
General recommendation is to use ViewModel for all UserContols. I guess this is true for those with more or less complex logic?

## What I am trying to do
I have a UserControl with TreeView in it.

1. Add ViewModel. The code below was the only one that would instantiate `vm:InspectorViewModel` automatically
```
    <UserControl.DataContext>
        <vm:InspectorViewModel />
    </UserControl.DataContext>
```

2. Add `TreeView`. At this point, everything works fine. `Entities` is a property of the `vm:InspectorViewModel`
```
<TreeView Name="EntityTree" DockPanel.Dock="Bottom" Items="{Binding Entities}">
    <TreeView.ItemTemplate>
        <TreeDataTemplate ItemsSource="{Binding Children}">
            <StackPanel Orientation="Horizontal">
                <PathIcon Data="{Binding Components, Converter={StaticResource EntityIconConverter}}" />
                <TextBlock Text="{Binding Name}"/>
             </StackPanel>
        </TreeDataTemplate>
    </TreeView.ItemTemplate>
</TreeView>
```

3. Now I want to handle `Tapped` event. The first option works but my assumption is that it is not recommended way:
```
<TreeView Name="EntityTree" DockPanel.Dock="Bottom" Items="{Binding Entities}" Tapped="EntityTree_OnTapped">
```
Where `EntityTree_OnTapped` is a method of `Inspector.axaml.cs` class but not `vm:InspectorViewModel`.

4. My assumption is that I need to create ReactiveCommand property in `vm:InspectorViewModel`? I that is the case, the only documentation I found (link above) tells me that In need to bind UserControl with `vm:InspectorViewModel`. And I did it this way:

```
public partial class Inspector : ReactiveUserControl<InspectorViewModel> {
  public TreeView EntityTreeView => this.FindControl<TreeView>("EntityTree");
  
  public Inspector() {
    this.WhenActivated(disposables => {
      Console.WriteLine($"Inspector activated {EntityTreeView}");
      // Bind the 'ExampleCommand' to 'ExampleButton' defined above.
    });
    AvaloniaXamlLoader.Load(this);
  }
...
```

Again, feels like a lot of boilerplate and snippet below works just fine:
```
    <UserControl.DataContext>
        <vm:InspectorViewModel />
    </UserControl.DataContext>
```

My assumption is that I am supposed to do:
```
this.WhenActivated(disposables => {
      this.BindCommand(
        ViewModel,
        vm => vm.OnTreeViewTappedCommand,
        self => self.EntityTreeView,
        "Tapped"
      ).DisposeWith(disposables);
});
```

Where `OnTreeViewTappedCommand`
```
public ICommand OnTreeViewTappedCommand { get; private set; }
...
this.WhenActivated(disposables => {
      Console.WriteLine($"InspectorViewModel activataed");
      
      OnTreeViewTappedCommand = ReactiveCommand.CreateFromObservable<object?, Unit>(OnTreeViewTapped);
      
      Disposable
        .Create(() => { /* handle deactivation */ })
        .DisposeWith(disposables);
    });

...

private static IObservable<Unit> OnTreeViewTapped(object? sender) {
  ...
}
```

This code is the only thing I was able to compile but app would crush at runtime if I Tap any TreeViewItem.

I would expect it because `OnTapped` event have `(object? sender, RoutedEventArgs e)` arguments

## What I am asking for
1. How to bind ViewModel to UserControl's TreeView Tap event? I found example only for Button's `click` event but Button have `Command` property which is used instead of `OnClick` event while TreeView does not have one.
2. Is there a simple way? Everything I posted above feels way to complex for simple `Tap` handling.
3. Is true that ideally, I should use my own ViewModel classes for all UserControl's? In that case every `control.axml.cs` file becomes just a boilerplate like:

```
public partial class Inspector : UserControl {
  public Properties() {
    InitializeComponent();
  }

  private void InitializeComponent() {
    AvaloniaXamlLoader.Load(this);
  }
}
```

Or boilerplate for binding Controls with associated ViewModel.

Thanks!

**Comment by maxkatz6** [comment, marked as answer]
> The first option works but my assumption is that it is not recommended way

Why would it be not recommended way? If you are asking about MVVM pattern, then it's fine to have a View related logic in the code behind and redirect call to the view model layer there. Though some projects might have stricter rules and not allow it.

> My assumption is that I need to create ReactiveCommand property in vm:InspectorViewModel

If you want to follow MVVM pattern, then yes, usually, you will have a command in the view model. And this command is executed from the view layer code (from XAML or C# code-behind).

> My assumption is that I am supposed to do:

Wiring View layer and ViewModel layer is very depended on specific MVVM framework what you are using. In case of ReactiveUI there are richer [documentation on their web site](https://www.reactiveui.net/docs/handbook/commands/binding-commands). Avalonia can be used with any MVVM framework that targets netstandard/net5 or higher.

> This code is the only thing I was able to compile but app would crush at runtime if I Tap any TreeViewItem.

Without knowing the exception hard to tell. But I assume you might be right about arguments.

> I found example only for Button's click event but Button have Command property which is used instead of OnClick event while TreeView does not have one.

Yes, Button.Command is the special one, because it's the most common. Also, it's integrated with Command.CanExecute, so button can be automatically disabled/enabled.

> Is there a simple way? Everything I posted above feels way to complex for simple Tap handling.

From my head:
1. Simplest. Add a C# event handler and execute a command from there:
```
// for ReactiveUI:
ViewModel.OnTreeViewTappedCommand.Execute(param).Subscribe();
```
2. Use [Behaviors SDK](https://github.com/wieslawsoltes/AvaloniaBehaviors) which allows to subscribe to events and execute commands on them in the XAML without using any C#. Many developers prefer this way.
3. Solve issue with BindCommand usage and parameter passing to the command. I am not so familiar with this way, so can't help here.

> Is true that ideally, I should use my own ViewModel classes for all UserControl's? In that case every control.axml.cs file becomes just a boilerplate like:

You can write as you want, but it's common to use MVVM pattern with XAML based applications for various reasons.
MVVM pattern forces you to separate View logic from the Model. Ideally, Model and ViewModel layers shouldn't know about View at all. And View layer shouldn't do any business logic not related to the View.
So, if your UserControl can be isolated from any business logic, it's fine to write a code behind code there (.axaml.cs file code) without any view model. Similarly, how controls like Button or TextBox are written - they can be reused and not depend on specific application. As you don't need a view model per each button obviously.
I would recommend finding books/tutorials about MVVM pattern to understand it better. Doesn't matter which framework you use, the point of the pattern is the same with Avalonia, WPF, UWP, Xamarin...

**Reply by Umenokin** [reply]
Thanks, @maxkatz6! I understand. I think understand MVVM conceptually but using it the first time, thus it is not always clear to me "the recommended way of doing things". I think `ViewModel.OnTreeViewTappedCommand.Execute(param).Subscribe();` sounds most reasonable to separate business logic, thus I will not need to expose TreeView-specific types to my ViewModel.
